services/dashboard/services/models/dag.js

```diff
@@ -0,0 +1,133 @@
+/*
+ * Licensed under the Apache License, Version 2.0
+ * See accompanying LICENSE file.
+ */
+
+angular.module('io.gearpump.models')
+
+  .service('Dag', function() {
+    'use strict';
+
+    /** This class represents a DAG. The topology is immutable after creation. */
+    function Dag(processors, edges) {
+      this.processors = processors;
+      this.edges = edges;
+      this.sortedProcessorIds = this._calculateTopologicalOrder();
+      this.predecessors = this._calculatePredecessors();
+      this.degrees = this._calculateDegrees();
+    }
+
+    Dag.prototype = {
+
+      isTopologyChanged: function(processors, edges) {
+        return !_.isEqual(this.processors, processors) || !_.isEqual(this.edges, edges);
+      },
+
+      getProcessorIds: function() {
+        return this.sortedProcessorIds;
+      },
+
+      getSourceProcessorIds: function() {
+        return _.filter(this.getProcessorIds(), function(processorId) {
+          return this.degrees[processorId].indegree === 0;
+        }, this);
+      },
+
+      getSinkProcessorIds: function() {
+        return _.filter(this.getProcessorIds(), function(processorId) {
+          return this.degrees[processorId].outdegree === 0;
+        }, this);
+      },
+
+      hierarchyDepth: function() {
+        return _.max(_.pluck(this.processors, 'hierarchy'));
+      },
+
+      _calculateTopologicalOrder: function() {
+        return _(this.processors).sortBy('hierarchy').pluck('id').value();
+      },
+
+      _calculatePredecessors: function() {
+        var result = {};
+        _.forEach(this.getProcessorIds(), function(processorId) {
+          result[processorId] = this._calculatePredecessorIds(processorId);
+        }, this);
+        return result;
+      },
+
+      _calculatePredecessorIds: function(processorId) {
+        var result = [];
+        _.forEach(this.edges, function(edge) {
+          if (edge.to === processorId) {
+            result.push(edge.from);
+          }
+        });
+        return result;
+      },
+
+      _calculateDegrees: function() {
+        var result = {};
+        _.forEach(this.processors, function(_, key) {
+          result[key] = {indegree: 0, outdegree: 0};
+        });
+
+        _.forEach(this.edges, function(edge) {
+          result[edge.from].outdegree++;
+          result[edge.to].indegree++;
+        });
+        return result;
+      },
+
+      /** Return the critical path. */
+      calculateCriticalPath: function(metricsProvider, time) {
```

Comment:
Yes. The javascript does not hint return type, if you are not using a *good* ide. So I'll describe return type of public methods, those method name does not imply its return type obviously.